packages/block-library/src/block/edit.js

```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { Placeholder, Spinner, Disabled } from '@wordpress/components';
-import { withSelect, withDispatch } from '@wordpress/data';
-import { __ } from '@wordpress/i18n';
 import {
 	BlockEditorProvider,
 	BlockList,
 	WritingFlow,
 } from '@wordpress/block-editor';
-import { compose } from '@wordpress/compose';
 import { parse, serialize } from '@wordpress/blocks';
+import { Placeholder, Spinner, Disabled } from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
 import ReusableBlockEditPanel from './edit-panel';
 
-class ReusableBlockEdit extends Component {
-	constructor( { reusableBlock } ) {
-		super( ...arguments );
-
-		this.startEditing = this.startEditing.bind( this );
-		this.stopEditing = this.stopEditing.bind( this );
-		this.setBlocks = this.setBlocks.bind( this );
-		this.setTitle = this.setTitle.bind( this );
-		this.save = this.save.bind( this );
-
-		if ( reusableBlock ) {
-			// Start in edit mode when we're working with a newly created reusable block
-			this.state = {
-				isEditing: reusableBlock.isTemporary,
-				title: reusableBlock.title,
-				blocks: parse( reusableBlock.content ),
-			};
-		} else {
-			// Start in preview mode when we're working with an existing reusable block
-			this.state = {
-				isEditing: false,
-				title: null,
-				blocks: [],
+const EMPTY_ARRAY = [];
+
+export default function ReusableBlockEdit( { attributes, isSelected } ) {
+	const { ref } = attributes;
+
+	const {
+		theReusableBlock,
+		isFetching,
+		isSaving,
+		isTemporary,
+		blocks,
+		canUpdateBlock,
+		settings,
+		title,
+	} = useSelect(
+		/** @typedef {import('@wordpress/blocks').BlockInstance} BlockInstance */
+
+		/**
+		 * @typedef WPReusableBlockEditSelectedData
+		 *
+		 * @property {?Array<BlockInstance>} blocks           Reusable block
+		 *                                                    content parsed as
+		 *                                                    blocks.
+		 * @property {boolean}               canUpdateBlock   Can the current
+		 *                                                    user edit reusable
+		 *                                                    blocks.
+		 * @property {boolean}               isFetching       Is the reusable
+		 *                                                    block being fetched.
+		 * @property {boolean}               isSaving         Is the reusable
+		 *                                                    block being saved.
+		 * @property {?boolean}              isTemporary      Is the reusable
+		 *                                                    block temporary.
```

Comment:
To my previous point, it's also not obvious if `isTemporary` should actually be nullable, or if this was just an oversight in the initial implementation. A consistent boolean type seems preferable, which we could change in the reducer / selector / wherever this is produced. To need to make that change and then return here to update these types is going to cause a lot of syncing errors / inconsistencies.

Reply:
Sometimes the reusable block is `null`, so it didn't seem right to always return a boolean for a value which is derived from an object that may or may not exist.

Reply:
To clarify, `isTemporary` can be `null`. I'm getting it using `_reusableBlock?.title ?? null`. I could default it to `false`, but I'm not sure if that's technically the right thing to do.

Reply:
My point here was more about: It may be true that the selector returns a `null` value, but it may very well be _wrong_ in doing so. We can fix that at the selector, but if we had kept these duplicate types here, it would be at high risk for falling out of sync.

Reply:
Now that the types are gone, do you think is fine for now?

Reply:
> Now that the types are gone, do you think this is fine for now?

Yes, the issue was mostly about the type definitions.